### DISCUSSION TITLE
(optional) significant whitespace for node children

### DISCUSSION THREAD

**Discussion by eugenesvk** [kdl-org/kdl#328] [General]
The FAQ correctly identifies one of the issues that YAML has with its significant whitespace

https://kdl.dev/#faq

> There's also a persistent issue where very large YAML files become unmanageable, especially due to the significant indentation feature

But then for **smaller** files significant whitespace beats all those empty lines with closing `}` taking too much space to signal indentation

What about **optional** significant whitespace for children? Then if your document is very large, you use `{` `}` just like now, and if not - then you can use indentation to signal the same info (or you have some other rule when to use what).

**Comment by tabatkins** [comment]
In general, you don't have quite as much nesting in a KDL document in the first place, due to the richer structure of a node. That is, rather than:

```json
{
  "type": "foo",
  "children": [
    {
      "type": "bar",
      ...
    },
    { 
      "type": "baz",
      ...
    }
  ]
}
```

You get the simpler:

```kdl
foo {
  bar ...
  baz ...
}
```

So the benefit of omitting the `}` lines isn't nearly as large as in JSON. And by the time it *is* significant, you've probably crossed the line into "large document" territory, where significant whitespace becomes a footgun.

**Reply by eugenesvk** [reply]
> So the benefit of omitting the `}` lines isn't nearly as large as in JSON

But even this repo's short example already illustrates the benefit quite nicely https://github.com/kdl-org/kdl/blob/main/examples/website.kdl

```
                    li "Ease of implementation"
                }
            }
        }
    }
}
```

what a waste of precious vertical space! And this is not a "large document" territory, so no footguns

**Reply by tabatkins** [reply]
That's at the very end of the document; a bunch of lines of closing parens aren't significant there. They're not pushing useful content further apart and making it harder to take in the document as a whole.

For the *rest* of the document, the `}`-only lines are relatively few and far between.

**Reply by eugenesvk** [reply]
> For the _rest_ of the document, the `}`-only lines are relatively few and far between

not really, this has the same issue
```
                    }
                    - " page in the Github repo. Feel free to jump in and give us your 2 cents!"
                }
            }
```

In this doc you waste ~20% of the height on empty lines, also `{` aren't a pretty sight either and not very convenient to type

**Comment by zkat** [comment]
I’m pretty strongly opposed to significant indentation. This is something I very intentionally left out because I consider it an incredibly painful part of working with YAML, where a slight misalignment in a large config file can lead to serious consequences. 

The aesthetics of a } cascade are insignificant in the face of genuine foot guns (and the added mental overhead of constantly having to double-check to protect yourself)

**Reply by eugenesvk** [reply]
But why are you strongly opposed to **optional** significant indentation? This is **not** meant for those huge files with footguns and overhead, but for the cleanest smaller config experience, where all the `{}` are just extra noise

**Reply by zkat** [reply]
Because I don’t want to complicate the grammar as much as possible, and I’d like a language that’s consistent in its  syntax style.
